### ISSUE TITLE
[Sandbox] krkn

### ISSUE THREAD

**Issue by psuriset** [cncf/sandbox#44]
### Application contact emails

nelluri@redhat.com , briordan@redhat.com, psuriset@redhat.com

### Project Summary

Krkn is a chaos and resiliency testing tool for Kubernetes with a focus on evaluating performance under load and scale

### Project Description

Krkn is a chaos and resiliency testing tool for Kubernetes with a focus on evaluating performance under load and scale.  It injects deliberate failures into the clusters to check if it is resilient to turbulent conditions and cluster as well as the components under test undergo minimal performance degradation to meet the SLOs and SLAs.
In addition to performance and resiliency evaluation, health of the entire cluster is also checked as disruption of one component might have an impact on other components.
Methodology is captured as part of the Chaos Testing Guide: https://redhat-chaos.github.io/krkn/. 

<img width="651" alt="Screenshot 2023-06-12 at 2 10 39 PM" src="https://github.com/cncf/sandbox/assets/13145336/900b4ac0-a532-4752-8bf5-a9ee7f0879bb">

### Org repo URL

https://github.com/redhat-chaos

### Project repo URL
All repos under in this org (https://github.com/redhat-chaos): 

Primary project is krkn  (https://github.com/redhat-chaos/krkn) & rest of them belong to krkn ecosystem. 

### Website URL

https://github.com/redhat-chaos/krkn and https://redhat-chaos.github.io/krkn/

### Roadmap

https://github.com/redhat-chaos/krkn/blob/main/ROADMAP.md

### Roadmap context

The roadmap items are prioritized based on the requests and feedback from the community. Each of the items added to the roadmap are thoroughly evaluated, discussed before adding the support in the tool.

### Contributing Guide

https://github.com/redhat-chaos/krkn/blob/main/docs/contribute.md 

### Code of Conduct (CoC)

https://github.com/redhat-chaos/krkn/blob/main/CODE_OF_CONDUCT.md

### Adopters

STARTX: French consulting and training company delivering expertise to build private and public cloud using Open-Source solutions. Red Hat: Engineering teams use it to chaos test Kubernetes, OpenShift and services running on top of it in CI environments like Prow. IBM Research: Engineering teams use it to develop chaos pipelines in software products - AI Framework integration to expand the capabilities after opening enhancements in Krkn and reviewed by maintainer and contributor Krkn team.

### Contributing or Sponsoring Org

Red Hat Inc.

### Maintainers file

https://github.com/redhat-chaos/krkn/blob/main/MAINTAINERS.md 

### IP Policy

- [X] If the project is accepted, I agree the project will follow the CNCF IP Policy

### Trademark and accounts

- [X] If the project is accepted, I agree to donate all project trademarks and accounts to the CNCF

### Why CNCF?

Our team firmly believes that Krkn is a valuable addition to the CNCF's impressive range of projects and could greatly benefit from support, oversight and direction that CNCF provides. The Krkn community would benefit from CNCF donation by allowing additional contributions and maintainers from the broader CNCF community. Krkn also provides features and capabilities that are not covered by other CNCF projects so the greater CNCF community would find Krkn’s adoption valuable as well.
In addition, Krkn being part of CNCF will help adopt good principles of governance and get guidance to build it into an incubation level project moving forward.

### Benefit to the Landscape

We would like to share the tooling and methodology of evaluating performance under chaotic conditions with the community involved in improving the performance, resilience and reliability of Kubernetes and services running on top of Kubernetes. As part of this, also maintain a chaos testing guide to help users get started: https://redhat-chaos.github.io/krkn/. 

This will help the ecosystem and community in cross-pollinating the areas of resilience/chaos and performance for Kubernetes.

### Cloud Native Overlap

Krkn share similar interests interms of chaos engineering with Litmus and Chaos-mesh but differs in terms of methodology especially considering performance as the main factor in addition to resilience.

Krkn provides the following capabilities that are not covered by Litmus or Chaos-mesh:

1. Designed keeping performance evaluation under chaotic conditions as the focus in addition to fault injection and recovery checks to make sure the component under test has minimal impact on the SLOs and SLAs.
2. Validates health of the entire cluster for pass/fail of a scenario as impact on one of the components can potentially disrupt other components.
3. Active work to support scenarios to run in parallel under load to mimic real world outages where multiple things can go wrong at once.
4. Runs outside the cluster to make sure it doesn’t become a victim of its own chaos running inside the cluster and reports the right results. 

### Similar projects

https://github.com/litmuschaos/litmus
https://github.com/chaos-mesh/chaos-mesh 

### Product or Service to Project separation

Krkn is an 100% open source project with a focus on community and not a product. Currently it’s under redhat-chaos github organization but will be a separate repository if accepted for CNCF sandboxing.

### Project presentations

Blog post on introduction to Kraken:[ https://www.openshift.com/blog/introduction-to-kraken-a-chaos-tool-for-openshift/kubernetes](https://www.openshift.com/blog/introduction-to-kraken-a-chaos-tool-for-openshift/kubernetes)
Discussion and demo on how Kraken can be leveraged to ensure Kubernetes/OpenShift is reliable, performant and scalable:[ https://www.youtube.com/watch?v=s1PvupI5sD0&ab_channel=OpenShift](https://www.youtube.com/watch?v=s1PvupI5sD0&ab_channel=OpenShift)
Blog post emphasizing the importance of making Chaos part of Performance and Scale runs to mimic the production environments:[ https://www.openshift.com/blog/making-chaos-part-of-kubernetes/openshift-performance-and-scalability-tests](https://www.openshift.com/blog/making-chaos-part-of-kubernetes/openshift-performance-and-scalability-tests)
Blog post on findings from Chaos test runs:[ https://cloud.redhat.com/blog/openshift/kubernetes-chaos-stories](https://cloud.redhat.com/blog/openshift/kubernetes-chaos-stories)

**Comment by joshgav** [comment]
@prithvi1307 @umamukkara @ksatchit as leads of <https://github.com/cncf/chaosengineering-wg> it would be great to get your thoughts here.

Let's try to include the krkn contributors in the work you're pursuing on chaos/resiliency testing generally. Thanks!

**Comment by AloisReitbauer** [comment]
Hello,

could you present the project in an upcoming TAG App Delivery meeting.

**Comment by psuriset** [comment]
@amye https://github.com/redhat-chaos/krkn is the repository in scope.
 It leverages https://github.com/redhat-chaos/cerberus for cluster level health checks which we are totally fine with sandboxing as well or integrate it as part of Krkn. Thanks.

**Comment by jberkus** [comment]
@psuriset none of the other repos?

**Comment by joshgav** [comment]
I'm excited to hear from krkn and @psuriset in TAG App Delivery and I believe testing and verification is indeed a part of application delivery. For example, [LitmusChaos](https://www.cncf.io/projects/litmus/) is managed by our TAG and there was even some effort to start a Chaos Engg WG with us.

But some testing frameworks like [ChaosBlade](https://www.cncf.io/projects/chaosblade/) and [ChaosMesh](https://www.cncf.io/projects/chaosmesh/) are managed by TAG Security. Let's be sure we're aligned on where test and verification frameworks fit in our landscape.

@psuriset let's get you on the TAG schedule, I think there's still some time at our next general meeting on Wednesday 9/20. Here's the agenda/notes doc, add yourself and I'll let the other leads know: <https://docs.google.com/document/d/1OykvqvhSG4AxEdmDMXilrupsX2n1qCSJUWwTc3I7AOs/edit>

Thanks!

**Comment by psuriset** [comment]
@jberkus Other repos are not in scope except the ones listed in this comment. https://github.com/cncf/sandbox/issues/44#issuecomment-1716007652

**Comment by amye** [comment]
The TOC was unclear about which repos were being contributed. Recommend that you re-organize your project repositories and more clearly indicate what's here, as well as editing your application so that it is clear.

**Comment by psuriset** [comment]
@amye  Apologies for the confusion. We re-organized our project repos in our github org. Everything under this org belongs to krkn ecosystem which we would like to move to cncf. We will give admin access to this org once application is approved. We can also rename it to be more community centric. https://github.com/redhat-chaos

Updated the application as well.

**Comment by joshgav** [comment]
Thank you @psuriset and team for presenting krkn to TAG App Delivery. Our writeup along with links to the recording and slides from the presentation are at <https://github.com/cncf/tag-app-delivery/issues/465#issuecomment-1776075439>.

The TAG believes krkn is a good fit for CNCF sandbox.

**Comment by TheFoxAtWork** [comment]
@psuriset Please reach out to TAG Security to ensure they are aware of this project. Several other Chaos testing tools are aligned under TAG Security. 

@joshgav would TAG App Delivery coordinate with TAG Security to clarify where chaos projects are homed?

**Comment by chaitanyaenr** [comment]
> @psuriset Please reach out to TAG Security to ensure they are aware of this project. Several other Chaos testing tools are aligned under TAG Security.
> 
> @joshgav would TAG App Delivery coordinate with TAG Security to clarify where chaos projects are homed?

@TheFoxAtWork Sure thing, we will work with @joshgav to share Krkn project details with the TAG Security. Thanks!

**Comment by joshgav** [comment]
Please note @mauilion's vote was made in <https://github.com/cncf/sandbox/issues/44#issuecomment-1858165359> instead of <https://github.com/cncf/sandbox/issues/44#issuecomment-1852449483> and so is not being counted.

With that vote 8/11 are in favor and this should be accepted to sandbox.

**Comment by psuriset** [comment]
Thank you accepting our project into CNCF Sandbox program.
